ambry-account/src/main/java/com/github/ambry/account/RouterStore.java

```diff
@@ -0,0 +1,331 @@
+    return router.putBlob(properties, null, channel, PutBlobOptions.DEFAULT).get();
+  }
+
+  private class ZKUpdater implements DataUpdater<ZNRecord> {
+    private final Collection<Account> accounts;
+    private final Pair<String, Path> backupPrefixAndPath;
+    private Map<String, String> potentialNewState;
+    private String newBlobID = null;
+
+    ZKUpdater(Collection<Account> accounts) {
+      this.accounts = accounts;
+      Pair<String, Path> backupPrefixAndPath = null;
+      if (accountService.backupDirPath != null) {
+        try {
+          backupPrefixAndPath = accountService.reserveBackupFile();
+        } catch (IOException e) {
+          logger.error("Error reserving backup file", e);
+        }
+      }
+      this.backupPrefixAndPath = backupPrefixAndPath;
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord znRecord) {
+      // There are several steps to finish an update
+      // 1. Fetch the list from the ZNRecord
+      // 2. Fetch the AccountMetadata from the blob id if the list exist in the ZNRecord
+      // 3. Construct a new AccountMetadata
+      // 4. save it as a blob in the ambry server
+      // 5. Add the new blob id back to the list.
+
+      // Start step 1:
+      ZNRecord recordToUpdate;
+      if (znRecord == null) {
+        logger.debug(
+            "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+            ACCOUNT_METADATA_BLOB_IDS_PATH);
+        recordToUpdate = new ZNRecord(ZN_RECORD_ID);
+      } else {
+        recordToUpdate = znRecord;
+      }
+
+      List<String> accountBlobIDs = recordToUpdate.getListField(ACCOUNT_METADATA_BLOB_IDS_LIST_KEY);
+      int newVersion = 1;
+      Map<String, String> accountMap = null;
+      if (accountBlobIDs != null && accountBlobIDs.size() != 0) {
+        // parse the json string list and get the blob id with the latest version
+        try {
+          BlobIDAndVersion blobIDAndVersion = null;
+          for (String accountBlobIDInJson : accountBlobIDs) {
+            BlobIDAndVersion current = BlobIDAndVersion.fromJson(accountBlobIDInJson);
+            if (blobIDAndVersion == null || blobIDAndVersion.version < current.version) {
+              blobIDAndVersion = current;
+            }
+          }
+          newVersion = blobIDAndVersion.version + 1;
+
+          // Start Step 2:
+          accountMap = readAccountMetadataFromBlobID(blobIDAndVersion.blobID);
+          // make this list mutable
+          accountBlobIDs = new ArrayList<>(accountBlobIDs);
+        } catch (JSONException e) {
+          logger.error("Exception occurred when parsing the blob id list from {}", accountBlobIDs);
+          accountService.accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+          throw new IllegalStateException("Exception occurred when parsing blob id list", e);
```

Comment:
why not use the same error message that you logged, and skip the logging?  If the caller will eventually log the exception, we don't want to log it multiple times on stack.

Reply:
Helix would actually catch the exception without logging it out. The function is basically a callback function and throwing any Exception would terminate the execution of HelixPropertyStore.update function and force it to return false.

Reply:
Fine.  Then create the string once with String.format("... %s") and use in both places.